### PR TITLE
DM-39915: Make butler.datastore an internal property

### PR DESCRIPTION
**Pull request by timj**
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/script/queryDatasets.py [hunk 1]
```diff
@@ -195,7 +195,7 @@ def getTables(self) -> list[AstropyTable]:
                 tables[dataset_ref.datasetType.name].add(dataset_ref)
         else:
             d = list(self.datasets)
-            ref_uris = self.butler.datastore.getManyURIs(d, predict=True)
+            ref_uris = self.butler._datastore.getManyURIs(d, predict=True)
```
Comment (timj) [inline python/lsst/daf/butler/script/queryDatasets.py]: I imagine we need to create a `butler.get_many_uris()` method and switch to that here.

File: python/lsst/daf/butler/_butler.py [hunk 2]
```diff
@@ -1265,8 +1265,8 @@ def getDirectDeferred(
         LookupError
             Raised if no matching dataset exists in the `Registry`.
         """
-        # Check thad dataset actuall exists.
-        if not self.datastore.exists(ref):
+        # Check thad dataset actually exists.
```
Comment (andy-slac) [inline python/lsst/daf/butler/_butler.py]:
```suggestion
        # Check that dataset actually exists.
```

File: python/lsst/daf/butler/_limited_butler.py [hunk 3]
```diff
@@ -274,6 +275,137 @@ def getDeferred(
...
+            If `True`, allow URIs to be returned of datasets that have not
+            been written.
+
+        Returns
+        -------
+        uris : `DatasetRefURIs`
+            The URI to the primary artifact associated with this dataset (if
+            the dataset was disassembled within the datastore this may be
+            `None`), and the URIs to any components associated with the dataset
+            artifact. (can be empty if there are no components).
```
Comment (andy-slac) [inline python/lsst/daf/butler/_limited_butler.py]:
```suggestion
            artifact (can be empty if there are no components).
```

File: python/lsst/daf/butler/_limited_butler.py [hunk 4]
```diff
@@ -274,6 +275,137 @@ def getDeferred(
...
+
+        Parameters
+        ----------
+        refs : iterable of `DatasetIdRef`
+            References to the required datasets.
+        predict : `bool`, optional
+            If the datastore does not know about a dataset, should it
+            return a predicted URI or not?
+        allow_missing : `bool`
+            If `False`, and `predict` is `False`, will raise if a `DatasetRef`
```
Comment (andy-slac) [inline python/lsst/daf/butler/_limited_butler.py]:
```suggestion
            If `False`, and ``predict`` is `False`, will raise if a `DatasetRef`
```

File: python/lsst/daf/butler/_limited_butler.py [hunk 5]
```diff
@@ -274,6 +275,137 @@ def getDeferred(
...
+    ) -> dict[DatasetRef, DatasetRefURIs]:
+        """Return URIs associated with many datasets.
+
+        Parameters
+        ----------
+        refs : iterable of `DatasetIdRef`
+            References to the required datasets.
+        predict : `bool`, optional
+            If the datastore does not know about a dataset, should it
+            return a predicted URI or not?
```
Comment (andy-slac) [inline python/lsst/daf/butler/_limited_butler.py]: Using a question for a description can raise some questions 🙂 Maybe reuse more assertive description as for `getURI`?

File: python/lsst/daf/butler/_limited_butler.py [hunk 6]
```diff
@@ -274,6 +275,137 @@ def getDeferred(
...
+        allow_missing : `bool`
+            If `False`, and `predict` is `False`, will raise if a `DatasetRef`
+            does not exist.
+
+        Returns
+        -------
+        URIs : `dict` of [`DatasetRef`, `DatasetRefUris`]
```
Comment (andy-slac) [inline python/lsst/daf/butler/_limited_butler.py]:
```suggestion
        URIs : `dict` of [`DatasetRef`, `DatasetRefURIs`]
```

File: python/lsst/daf/butler/_quantum_backed.py [hunk 7]
```diff
@@ -227,7 +227,7 @@ def initialize(
             predicted_outputs=predicted_outputs,
             dimensions=dimensions,
             filename=filename,
-            datastore_records=quantum.datastore_records,
+            datastore_records=quantum._datastore_records,
```
Comment (andy-slac) [inline python/lsst/daf/butler/_quantum_backed.py]: I think we should keep `datastore_records` here.

File: python/lsst/daf/butler/core/datastore.py [hunk 8]
```diff
@@ -372,6 +372,18 @@ def names(self) -> tuple[str, ...]:
         # Default implementation returns solely the name itself
         return (self.name,)
 
+    @property
+    def roots(self) -> dict[str, ResourcePath | None]:
+        """Return the root URIs for each named datastore.
+
+        Returns
```
Comment (andy-slac) [inline python/lsst/daf/butler/core/datastore.py]: Do we use `Returns` for `@property`? I thought we just write property type in parens.

File: python/lsst/daf/butler/script/queryDatasets.py (hunk 1 above)
Comment (andy-slac) [reply python/lsst/daf/butler/script/queryDatasets.py]: You have added that method, may use it here as well?

File: python/lsst/daf/butler/_limited_butler.py [hunk 9]
```diff
@@ -274,6 +275,137 @@ def getDeferred(
...
+
+    def get_datastore_roots(self) -> dict[str, ResourcePath | None]:
+        """Return the defined root URIs for all registered datastores.
+
+        Returns
+        -------
+        roots : `dict` [`str`, `~lsst.resources.ResourcePath` | `None`]
+            A mapping from datastore name to datastore root URI. The root
+            can be `None` if the datastore does not have any concept of a root
+            URI.
```
Comment (andy-slac) [inline python/lsst/daf/butler/_limited_butler.py:296]: Abstract question - I suppose it does not make sense to have a single datastore with multiple roots?

**Review by andy-slac — APPROVED** [review]
Looks good, few minor comments. I guess you need to figure out why build fails and mypy shows some new errors.

File: python/lsst/daf/butler/_quantum_backed.py (hunk 7 above)
Comment (timj) [reply python/lsst/daf/butler/_quantum_backed.py]: Hmm. It seems I fluked the internal name. That explains why the tests worked...

File: python/lsst/daf/butler/core/datastore.py (hunk 8 above)
Comment (timj) [reply python/lsst/daf/butler/core/datastore.py]: I wrote it as a getter first and then changed my mind.

File: python/lsst/daf/butler/_limited_butler.py (hunk 9 above)
Comment (timj) [reply python/lsst/daf/butler/_limited_butler.py:296]: Not the way we've currently set things up. Multiple roots means chained datastore at the moment.

**Comment by timj** [comment]
The build failing is from fastapi supporting pydantic v2. The mypy failure is all pydantic and we have started a discussion on Slack about it -- it won't block the merge.

**Comment by andy-slac** [comment]
> The mypy failure is all pydantic and we have started a discussion on Slack about it -- it won't block the merge.

I see GA mypy failures in `cli/utils.py` and `cli/cmd/commands.py` which I did not see before, maybe new `click` version?

**Comment by andy-slac** [comment]
> I see GA mypy failures in `cli/utils.py` and `cli/cmd/commands.py` which I did not see before, maybe new `click` version?

Indeed, I see the same errors locally if I upgrade click from 8.1.3 to 8.1.4.

**Comment by timj** [comment]
> I see the same errors locally if I upgrade click from 8.1.3 to 8.1.4.

which shows the dangers of having a failing action that we then ignore. I hadn't noticed the click issue.

**Pull request merged**